torchci/lib/metricUtils.ts

```diff
@@ -0,0 +1,193 @@
+  // The data from Rockset is sorted by time DESC, so newer commits come first
+  data?: JobsPerCommitData[],
+  broken_trunk_threshold: number = BROKEN_TRUNK_THRESHOLD,
+  outage_threshold: number = OUTAGE_THRESHOLD
+) {
+  const failuresByTypes: { [failure: string]: { [t: string]: number } } = {};
+
+  if (data === undefined || data === null) {
+    return failuresByTypes;
+  }
+
+  const failuresCount: { [failure: string]: number } = {};
+  data.forEach((commit: JobsPerCommitData) => {
+    const failures = new Set(
+      commit.failures.filter(
+        (n) => n !== null && n !== undefined && n.length > 0
+      )
+    );
+
+    // Iterate though all the failures in the commit and aggregate them by name
+    failures.forEach((failure: string) => {
+      if (!(failure in failuresCount)) {
+        // Start the count
+        failuresCount[failure] = 0;
+      }
+
+      failuresCount[failure] += 1;
+    });
+
+    // Check if the job still fail in this commit
+    Object.keys(failuresCount).forEach((failure: string) => {
+      if (failures.has(failure)) {
+        // Count the commit as part of an outage
+        if (failures.size >= outage_threshold) {
+          updateBrokenInfraCount(failure, 1, failuresByTypes);
+        }
+
+        // Still failing, its counter has already been updated
+        return;
+      }
+
+      const count = failuresCount[failure];
+      // Reaching here means that the job starts to fail on the commit after this
+      updateFailureCount(
+        failure,
+        count,
+        failuresByTypes,
+        count >= broken_trunk_threshold
+      );
+
+      // Reset the count
+      failuresCount[failure] = 0;
+    });
+  });
+
+  Object.keys(failuresCount).forEach((failure: string) => {
+    const count = failuresCount[failure];
+    // Aggregate all remaining jobs
+    updateFailureCount(
+      failure,
+      count,
+      failuresByTypes,
+      count >= broken_trunk_threshold
+    );
+  });
+
+  return failuresByTypes;
+}
+
+export function approximateFailureByTypePercent(
+  // The data from Rockset is sorted by time DESC, so newer commits come first
+  data?: JobsPerCommitData[],
+  broken_trunk_threshold: number = BROKEN_TRUNK_THRESHOLD,
+  outage_threshold: number = OUTAGE_THRESHOLD
+) {
+  const failuresByTypes = approximateFailureByType(
+    data,
+    broken_trunk_threshold,
+    outage_threshold
+  );
+  if (data === undefined || data === null || data.length === 0) {
+    return failuresByTypes;
+  }
+
+  // Get the number of times the job successes too, so we can calculate the %
+  const successesByJobName = approximateSuccessByJobName(data);
+
+  Object.keys(failuresByTypes).forEach((jobName: string) => {
+    const successCount = successesByJobName[jobName] ?? 0;
+    const failureCount = Object.entries(failuresByTypes[jobName])
+      .filter(
+        (item) =>
+          item[0] === JobAnnotation.BROKEN_TRUNK ||
+          item[0] === JobAnnotation.TEST_FLAKE
```

Comment:
Shouldn't outages be included here?

Reply:
Somehow I know someone would notice this.  At the moment, the outage count is estimated separately using the `total number of failures in a commit > the threshold of 10` heuristic.  So all commits satisfying that condition would be counted in that bucket.  You can see that it's not overlapping with flaky and broken trunk failures.  

I was rather reluctant to include that value at first because all I want is to divide the failure % into two groups 1) broken trunk and 2) non broken trunk (flaky).  In the end, I include the outage estimate to see how useful it is.  Let's think of it as a separate metric for the % of commits having many failed jobs (> 10).  Another way to display this information is to have a histogram of the number of failed jobs per commit